File: gfg-weekly-coding-contest-109/Friends Meet Up.cpp
//User function Template for C++

class Solution{
    public:
    // Version 1: Greedy
    // Sort by start time, and choose the earliest deadline friends
    // TC: O(nlogn + max(end) * logn), SC: O(n)
    int friends(int N, int K, vector<int> start, vector<int> end){
        // code here
        vector<pair<int, int>> f;
        for (int i = 0; i < N; i++) {
            f.push_back({start[i], end[i]});
        }
        sort(f.begin(), f.end());
        int ans = 0;
        int current = f[0].first;
        int i = 0;
        priority_queue<int, vector<int>, greater<int>> pq;
        while (i < N || !pq.empty()) {
            while (i < N && f[i].first == current) {
                pq.push(f[i].second);
                i++;
            }
            while (!pq.empty() && pq.top() < current) {
                pq.pop();
            }
            
            int count = 0;
            while (!pq.empty() && count < K) {
                ans++;
                count++;
                pq.pop();
            }
            current++;
        }
        return ans;
    }
};

File: gfg-weekly-coding-contest-109/Friends Meet Up.java


//User function Template for Java
class Solution 
{ 
    // Version 1: Greedy
    // Sort by start time, and choose the earliest deadline friends
    // TC: O(nlogn + max(end) * logn), SC: O(n)
    int friends(int N, int K, int start[] ,int end[]) 
    { 
        // code here
        ArrayList<ArrayList<Integer>> f = new ArrayList<ArrayList<Integer>>();
        for (int i = 0; i < N; i++) {
            f.add(new ArrayList<>());
            f.get(i).add(start[i]);
            f.get(i).add(end[i]);
        }
        Collections.sort(f, (l1, l2) -> Integer.compare(l1.get(0), l2.get(0)));
        
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>();
        int i = 0;
        int current = f.get(0).get(0);
        int ans = 0;
        while (i < f.size() || !pq.isEmpty()) {
            while (i < N && f.get(i).get(0) == current) {
                pq.offer(f.get(i).get(1));
                i++;
            }
            
            while (!pq.isEmpty() && pq.peek() < current) {
                pq.poll();
            }
            
            int count = 0;
            while (!pq.isEmpty() && count < K) {
                pq.poll();
                count++;
                ans++;
            }
            current++;
        }
        return ans;
    }
} 